File: packages/core/src/actions/ActionRouter.test.ts
import { describe, it, expect } from "vitest";
import { ActionRouter } from "./ActionRouter";
import { createAgentAction } from "./AgentAction";

describe("ActionRouter", () => {
  it("should execute allowed actions", async () => {
    const router = new ActionRouter();
    router.registerHandler("create_plan", async () => "Plan created");

    const action = createAgentAction(
      "create_plan",
      "Create Plan",
      "Create a new plan"
    );

    const result = await router.execute(action);
    expect(result.success).toBe(true);
    expect(result.result).toBe("Plan created");
  });

  it("should reject actions without handlers", async () => {
    const router = new ActionRouter();
    const action = createAgentAction("custom", "Custom", "Custom action");

    const result = await router.execute(action);
    expect(result.success).toBe(false);
    expect(result.error).toContain("No handler registered");
  });

  it("should require confirmation for destructive actions", async () => {
    const router = new ActionRouter();
    const action = createAgentAction(
      "create_plan",
      "Delete All",
      "Delete everything",
      {},
      false,
      true
    );

    const result = await router.execute(action);
    expect(result.success).toBe(false);
    expect(result.error).toContain("requires confirmation");
  });

  it("should track action history", async () => {
    const router = new ActionRouter();
    router.registerHandler("create_plan", async () => "Done");

    const action = createAgentAction(
      "create_plan",
      "Test",
      "Test action"
    );

    await router.execute(action);
    const history = router.getHistory();
    expect(history.length).toBe(1);
  });

  it("should handle action errors gracefully", async () => {
    const router = new ActionRouter();
    router.registerHandler("create_plan", async () => {
      throw new Error("Action failed");
    });

    const action = createAgentAction(
      "create_plan",
      "Test",
      "Test action"
    );

    const result = await router.execute(action);
    expect(result.success).toBe(false);
    expect(result.error).toBe("Action failed");
  });
});

File: packages/compose/src/test/kotlin/ai/mobileagentos/compose/ActionRouterTest.kt
package ai.mobileagentos.compose

import ai.mobileagentos.compose.actions.ActionRouter
import ai.mobileagentos.compose.actions.AgentAction
import kotlinx.coroutines.test.runTest
import org.junit.Assert.*
import org.junit.Test

class ActionRouterTest {
    
    @Test
    fun `execute action with handler`() = runTest {
        val router = ActionRouter()
        router.registerHandler("test_action") { action ->
            "Action ${action.type} executed"
        }
        
        val action = AgentAction.create("test_action")
        val result = router.execute(action)
        
        assertTrue(result.success)
        assertNotNull(result.result)
        assertTrue(result.result.toString().contains("test_action"))
    }
    
    @Test
    fun `execute action without handler fails`() = runTest {
        val router = ActionRouter()
        
        val action = AgentAction.create("unknown_action")
        val result = router.execute(action)
        
        assertFalse(result.success)
        assertNotNull(result.error)
        assertTrue(result.error!!.contains("No handler"))
    }
    
    @Test
    fun `action history is tracked`() = runTest {
        val router = ActionRouter()
        router.registerHandler("action1") { "Result 1" }
        router.registerHandler("action2") { "Result 2" }
        
        router.execute(AgentAction.create("action1"))
        router.execute(AgentAction.create("action2"))
        
        val history = router.getHistory()
        assertEquals(2, history.size)
    }
    
    @Test
    fun `clear history works`() = runTest {
        val router = ActionRouter()
        router.registerHandler("test") { "Result" }
        
        router.execute(AgentAction.create("test"))
        assertEquals(1, router.getHistory().size)
        
        router.clearHistory()
        assertEquals(0, router.getHistory().size)
    }
}